Add more quirks for `consent.google.*` when applying the third party IP address cookie expiry cap

whsieh opened the pull request:
#### 2f857be7d42f00feb7b59870709310154181a774
<pre>
Add more quirks for `consent.google.*` when applying the third party IP address cookie expiry cap
<a href="https://bugs.webkit.org/show_bug.cgi?id=271952">https://bugs.webkit.org/show_bug.cgi?id=271952</a>
<a href="https://rdar.apple.com/122811153">rdar://122811153</a>

Reviewed by John Wilander.

Add a couple more quirks for requests to `consent.google.*` and `consent.youtube.com`, when checking
whether we should apply the cookie expiration cap for requests coming from potential third party IP
addresses.

* Source/WebKit/NetworkProcess/cocoa/NetworkTaskCocoa.mm:
(WebKit::NetworkTaskCocoa::applyCookiePolicyForThirdPartyCloaking):

Clean up this capture block, so that the individual captured variables are much easier to read.

Canonical link: <a href="https://commits.webkit.org/276888@main">https://commits.webkit.org/276888@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/627a9b77205e45a741730e395d0e5ee3c89b2976

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/46057 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/25187 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/48642 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/48728 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/42097 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/29503 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/22624 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/37662 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/46635 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/22277 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/39738 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/18852 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/19676 "Passed tests") | | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/4101 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/42375 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/41185 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/50519 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/21053 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/17571 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/44825 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/22353 "Built successfully") | | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/43718 "Found 1 new API test failure: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/event-listener (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10202 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/22712 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/22047 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

johnwilander commented on Source/WebKit/NetworkProcess/cocoa/NetworkTaskCocoa.mm:
```diff
@@ -135,15 +135,23 @@ static bool shouldCapCookieExpiryForThirdPartyIPAddress(const WebCore::IPAddress
+        , thirdPartyCNAMEDomainForTesting = crossThreadCopy(m_networkSession->thirdPartyCNAMEDomainForTesting())
+        , ageCapForCNAMECloakedCookies = crossThreadCopy(m_ageCapForCNAMECloakedCookies)
+        , weakTask = WeakObjCPtr<NSURLSessionTask>(task())
+        , topDomainName = crossThreadCopy(RegistrableDomain { firstPartyURL }.string())
```
Let's call it `topRegistrableDomain` instead. See comment further down.

johnwilander replied:
Or `firstPartyRegistrableDomain` to match the others.

whsieh replied:
Good point! I'll go with `firstPartyRegistrableDomainName`.

johnwilander commented on Source/WebKit/NetworkProcess/cocoa/NetworkTaskCocoa.mm:
```diff
@@ -164,14 +172,19 @@ static bool shouldCapCookieExpiryForThirdPartyIPAddress(const WebCore::IPAddress
+                if (hostName == "accounts.google.com"_s)
+                    return true;
+
+                return (topDomainName.startsWith("google."_s) || topDomainName == "youtube.com"_s)
```
Down here, one worries that any site could add a `google.` subdomain to trigger this rule, for instance `google.some.unrelated.example.com`. However, with variable name `topRegistrableDomain` it will be clear that the rule is much more narrow than that.

Still, we need to acknowledge that there may be registrable domains starting with `google.` that don't belong to Google. You can imagine any entry on the Public Suffix List with a leading `google.`. For a soft mitigation like capped cookie lifetime, I think that's OK to keep this speedy. But it's important that harder policies don't rely on this overly relaxed rule.

johnwilander reviewed (approved):
I'm assuming you've tested this. :)

whsieh committed:
Add more quirks for `consent.google.*` when applying the third party IP address cookie expiry cap
https://bugs.webkit.org/show_bug.cgi?id=271952
rdar://122811153

Reviewed by John Wilander.

Add a couple more quirks for requests to `consent.google.*` and `consent.youtube.com`, when checking
whether we should apply the cookie expiration cap for requests coming from potential third party IP
addresses.

* Source/WebKit/NetworkProcess/cocoa/NetworkTaskCocoa.mm:
(WebKit::NetworkTaskCocoa::applyCookiePolicyForThirdPartyCloaking):

Clean up this capture block, so that the individual captured variables are much easier to read.

Canonical link: https://commits.webkit.org/276888@main
